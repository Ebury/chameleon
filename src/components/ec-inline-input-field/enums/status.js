export const EDITING = 'editing';
export const LOADING = 'loading';
export const READ_ONLY = 'read-only';
