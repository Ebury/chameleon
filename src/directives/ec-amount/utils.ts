
import defaultOptions from './options';
import type { AmountDirectiveOptions } from './types';

function format(input: undefined, opts?: AmountDirectiveOptions): undefined;
function format(input: null, opts?: AmountDirectiveOptions): null;
function format(input: string | number, opts?: AmountDirectiveOptions): string;
function format(input: string | number | undefined | null, opts: AmountDirectiveOptions = defaultOptions): string | null | undefined {
  /* eslint-disable no-param-reassign */
  if (input === '' || input === null || input === undefined) {
    return input;
  }

  if (typeof input === 'number') {
    input = input.toString();
  }

  const negative = input.startsWith('-') ? '-' : '';

  input = sanitizeInput(input, opts.decimalSeparator);
  if (opts.precision > 0) {
    input = keepOnlyOneDecimalSeparator(input, opts.decimalSeparator);
  } else {
    input = removeDecimal(input, opts.decimalSeparator);
  }
  const parts = input.split(opts.decimalSeparator);
  let integralPart = parseToStr(parts[0]);
  let fractionalPart = parts[1];

  if (fractionalPart && fractionalPart.length > opts.precision) {
    fractionalPart = fractionalPart.substring(0, opts.precision);
  }

  integralPart = addGroupingSeparators(integralPart, opts.groupingSeparator);
  const result = negative + joinIntegralAndFractionalParts(integralPart, fractionalPart, opts.decimalSeparator);
  return result;
}

function parseToStr(number: string): string {
  const parsed = Number.parseFloat(number);
  if (Number.isNaN(parsed)) {
    return '';
  }
  return parsed.toString();
}

function keepOnlyOneDecimalSeparator(number: string, decimalSeparator: string): string {
  if (number.includes(decimalSeparator)) {
    const split = number.split(decimalSeparator);
    return split[0] + decimalSeparator + split.slice(1).join('');
  }
  return number;
}

function removeDecimal(number: string, decimalSeparator: string): string {
  const split = number.split(decimalSeparator);
  return split[0];
}

function unFormat(input: string, groupingSeparator: string, decimalSeparator: string): string {
  if (!input) {
    return input;
  }

  return input
    .replace(new RegExp(`\\${groupingSeparator}`, 'g'), '')
    .replace(new RegExp(`\\${decimalSeparator}`, 'g'), '.');
}

function sanitizeInput(input: string, decimalSeparator: string): string {
  return input.replace(new RegExp(`[^0-9${decimalSeparator}]`, 'g'), '') || '';
}

function addGroupingSeparators(integralPart: string, separator: string): string {
  return integralPart.replace(/(\d)(?=(?:\d{3})+\b)/gm, `$1${separator}`);
}

function joinIntegralAndFractionalParts(integralPart: string, fractionalPart: string, decimalSeparator: string): string {
  return fractionalPart !== undefined ? integralPart + decimalSeparator + fractionalPart : integralPart;
}

/* c8 ignore start */
function setCursor(el: HTMLInputElement, position: number) {
  const setSelectionRange = function setSelectionRange() {
    el.setSelectionRange(position, position);
  };

  if (el === document.activeElement) {
    setSelectionRange();
    setTimeout(setSelectionRange, 1); // Android Fix
  }
}
/* c8 ignore stop */

export {
  format,
  sanitizeInput,
  setCursor,
  unFormat,
};
