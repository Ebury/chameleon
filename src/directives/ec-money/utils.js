import defaults from './options';

function format(input, opt = defaults) {
  if (input === '' || input === null || input === undefined) {
    return input;
  }

  if (typeof input === 'number') {
    input = input.toString();
  }

  const negative = input.includes('-') ? '-' : '';

  input = sanitizeInput(input, opt.decimalSeparator);
  input = keepOnlyOneDecimalSeparator(input, opt.decimalSeparator);
  const parts = input.split(opt.decimalSeparator);
  let integer = parseToStr(parts[0]);
  let decimal = parts[1];

  if (decimal && decimal.length > opt.precision) {
    decimal = decimal.substr(0, opt.precision);
  }

  integer = addGroupingSeparators(integer, opt.groupingSeparator);
  const result = negative + joinIntegerAndDecimal(integer, decimal, opt.decimalSeparator);
  return result;
}

function parseToStr(number) {
  /* eslint no-restricted-globals: "off" */
  const parsed = parseFloat(number);
  if (isNaN(parsed)) {
    return '';
  }
  return parsed.toString();
}

function keepOnlyOneDecimalSeparator(number, decimalSeparator) {
  if (number.includes(decimalSeparator)) {
    const split = number.split(decimalSeparator);
    return split[0] + decimalSeparator + split.slice(1).join('');
  }
  return number;
}

function unFormat(input, groupingSeparator, decimalSeparator) {
  /* eslint no-param-reassign: "off" */
  input = (`${input}`).replace(new RegExp(`\\${groupingSeparator}`, 'g'), '');
  input = (`${input}`).replace(new RegExp(`\\${decimalSeparator}`, 'g'), '.');
  return input;
}

function sanitizeInput(input, decimalSeparator) {
  return input.replace(new RegExp(`[^0-9${decimalSeparator}]`, 'g'), '') || '';
}

function addGroupingSeparators(integer, separator) {
  return integer.replace(/(\d)(?=(?:\d{3})+\b)/gm, `$1${separator}`);
}

function joinIntegerAndDecimal(integer, decimal, decimalSeparator) {
  return decimal !== undefined ? integer + decimalSeparator + decimal : integer;
}

function setCursor(el, position) {
  const setSelectionRange = function setSelectionRange() {
    el.setSelectionRange(position, position);
  };

  if (el === document.activeElement) {
    setSelectionRange();
    setTimeout(setSelectionRange, 1); // Android Fix
  }
}

// https://developer.mozilla.org/en-US/docs/Web/Guide/Events/Creating_and_triggering_events#The_old-fashioned_way
function event(name) {
  const evt = document.createEvent('Event');
  evt.initEvent(name, true, true);
  return evt;
}

export {
  format,
  unFormat,
  setCursor,
  event,
  sanitizeInput,
};
