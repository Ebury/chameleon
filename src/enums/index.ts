export enum SortDirection {
  ASC = 'asc',
  DESC = 'desc',
}

export enum SortDirectionCycle {
  LOWEST_FIRST = 1,
  HIGHEST_FIRST = 2,
}

export enum KeyCode {
  ESC = 27,
  ARROW_UP = 38,
  ARROW_DOWN = 40,
}

export enum ZIndexLevel {
  NOTIFICATION = 'notification',
  MODAL = 'modal',
  TOOLTIP = 'tooltip',
  LEVEL1 = 'level-1',
  LEVEL2 = 'level-2',
  LEVEL3 = 'level-3',
}
