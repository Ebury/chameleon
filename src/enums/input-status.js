export const EDIT = 'edit';
export const LOADING = 'loading';
export const READ_ONLY = 'read-only';
