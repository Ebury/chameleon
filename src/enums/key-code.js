export const ESC = 27;
