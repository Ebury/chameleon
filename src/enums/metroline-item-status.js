export const NEXT = 'next';
export const ACTIVE = 'active';
export const COMPLETED = 'completed';
