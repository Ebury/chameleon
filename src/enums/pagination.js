export const PAGE_SIZES = [5, 10, 50, 100];
export const DEFAULT_PAGE_SIZE = 10;
