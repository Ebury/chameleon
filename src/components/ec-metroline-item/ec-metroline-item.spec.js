import { mount } from '@vue/test-utils';
import { withMockedConsole } from '../../../tests/utils/console';
import EcMetrolineItem from './ec-metroline-item.vue';
import * as STATUS from '../../enums/metroline-status';

function mountEcMetrolineItem(props, mountOpts) {
  return mount(EcMetrolineItem, {
    propsData: {
      itemIndex: '1',
      ...props,
    },
    ...mountOpts,
  });
}

const slots = {
  heading: '<span>Payee information</span>',
  'sub-heading': '<span>Amount fully allocated</span>',
  'header-cta': '<a href="#>Edit</a>',
  main: '<p> Lorem ipsum dolor sit amet, voluptates at cumque repudiandae atque quod voluptatum, aperiam dignissimos, vitae, neque mollitia repellat! </p>',
  'footer-cta': '<a href="#>Continue</a>',
};

describe('EcMetrolineItem', () => {
  it('should throw an error when we pass a non valid status', () => {
    withMockedConsole((errorSpy) => {
      mountEcMetrolineItem({ status: 'a non valid status' });
      expect(errorSpy).toHaveBeenCalledTimes(1);
      expect(errorSpy.mock.calls[0][0]).toContain('Invalid prop: custom validator check failed for prop "status"');
    });
  });
  it('it should render properly when status is "next"', () => {
    const wrapper = mountEcMetrolineItem(
      {
        status: STATUS.NEXT,
      },
      {
        slots,
      },
    );
    expect(wrapper.findByDataTest('ec-metroline-item__status-disk-enumeration').exists()).toBe(true);
    expect(wrapper.findByDataTest('ec-metroline-item__status-disk-completed').exists()).toBe(false);
    expect(wrapper.findByDataTest('ec-metroline-item__header-heading').exists()).toBe(true);
    expect(wrapper.findByDataTest('ec-metroline-item__header-sub-heading').exists()).toBe(false);
    expect(wrapper.findByDataTest('ec-metroline-item__status-bar').exists()).toBe(true);
    expect(wrapper.findByDataTest('ec-metroline-item__header-cta').exists()).toBe(false);
    expect(wrapper.findByDataTest('ec-metroline-item__main').exists()).toBe(false);
    expect(wrapper.findByDataTest('ec-metroline-item__footer').exists()).toBe(false);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render properly when status is "active"', () => {
    const wrapper = mountEcMetrolineItem(
      {
        status: STATUS.ACTIVE,
      },
      {
        slots,
      },
    );

    expect(wrapper.findByDataTest('ec-metroline-item__status-disk-enumeration').exists()).toBe(true);
    expect(wrapper.findByDataTest('ec-metroline-item__status-disk-completed').exists()).toBe(false);
    expect(wrapper.findByDataTest('ec-metroline-item__header-heading').exists()).toBe(true);
    expect(wrapper.findByDataTest('ec-metroline-item__header-sub-heading').exists()).toBe(false);
    expect(wrapper.findByDataTest('ec-metroline-item__status-bar').exists()).toBe(true);
    expect(wrapper.findByDataTest('ec-metroline-item__header-cta').exists()).toBe(false);
    expect(wrapper.findByDataTest('ec-metroline-item__main').exists()).toBe(true);
    expect(wrapper.findByDataTest('ec-metroline-item__footer').exists()).toBe(true);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should properly when status is "completed"', () => {
    const wrapper = mountEcMetrolineItem(
      {
        status: STATUS.COMPLETED,
      },
      {
        slots,
      },
    );

    expect(wrapper.findByDataTest('ec-metroline-item__status-disk-enumeration').exists()).toBe(false);
    expect(wrapper.findByDataTest('ec-metroline-item__status-disk-completed').exists()).toBe(true);
    expect(wrapper.findByDataTest('ec-metroline-item__header-heading').exists()).toBe(true);
    expect(wrapper.findByDataTest('ec-metroline-item__header-sub-heading').exists()).toBe(true);
    expect(wrapper.findByDataTest('ec-metroline-item__status-bar').exists()).toBe(true);
    expect(wrapper.findByDataTest('ec-metroline-item__header-cta').exists()).toBe(true);
    expect(wrapper.findByDataTest('ec-metroline-item__main').exists()).toBe(true);
    expect(wrapper.findByDataTest('ec-metroline-item__footer').exists()).toBe(true);
    expect(wrapper.element).toMatchSnapshot();
  });

  it.each([STATUS.NEXT, STATUS.ACTIVE, STATUS.COMPLETED])('should render properly when is the last item and status is "%s"', (status) => {
    const wrapper = mountEcMetrolineItem(
      {
        status,
        isLastItem: true,
      },
      {
        slots,
      },
    );

    expect(wrapper.findByDataTest('ec-metroline-item__status-bar').exists()).toBe(false);
    expect(wrapper.element).toMatchSnapshot();
  });
});
