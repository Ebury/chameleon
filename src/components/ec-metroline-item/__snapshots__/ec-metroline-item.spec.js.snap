// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EcMetrolineItem it should render properly when status is "next" 1`] = `
<div
  class="ec-metroline-item"
  data-test="ec-metroline-item"
>
  <div
    class="ec-metroline-item__status"
    data-test="ec-metroline-item__status"
  >
    <div
      class="ec-metroline-item__badge"
      data-test="ec-metroline-item__badge"
    >
      <span
        data-test="ec-metroline-item__index"
      >
        
        1
      
      </span>
    </div>
     
    <div
      class="ec-metroline-item__status-bar"
      data-test="ec-metroline-item__status-bar"
    />
  </div>
   
  <div
    class="ec-metroline-item__content"
    data-test="ec-metroline-item__content"
  >
    <div
      class="ec-metroline-item__header"
      data-test="ec-metroline-item__header"
    >
      <div
        class="ec-metroline-item__header-heading"
        data-test="ec-metroline-item__header-heading"
      >
        <span>
          Payee information
        </span>
      </div>
       
      <!---->
       
      <!---->
    </div>
     
    <!---->
     
    <!---->
  </div>
</div>
`;

exports[`EcMetrolineItem should properly when status is "completed" 1`] = `
<div
  class="ec-metroline-item"
  data-test="ec-metroline-item"
>
  <div
    class="ec-metroline-item__status"
    data-test="ec-metroline-item__status"
  >
    <div
      class="ec-metroline-item__badge"
      data-test="ec-metroline-item__badge"
    >
      <svg
        class="ec-icon"
        data-test="ec-metroline-item__completed-icon"
        height="14"
        width="14"
      >
        <use
          xlink:href="#ec-simple-check"
        />
      </svg>
    </div>
     
    <div
      class="ec-metroline-item__status-bar"
      data-test="ec-metroline-item__status-bar"
    />
  </div>
   
  <div
    class="ec-metroline-item__content"
    data-test="ec-metroline-item__content"
  >
    <div
      class="ec-metroline-item__header"
      data-test="ec-metroline-item__header"
    >
      <div
        class="ec-metroline-item__header-heading"
        data-test="ec-metroline-item__header-heading"
      >
        <span>
          Payee information
        </span>
      </div>
       
      <div
        class="ec-metroline-item__header-sub-heading"
        data-test="ec-metroline-item__header-sub-heading"
      >
        <span>
          Amount fully allocated
        </span>
      </div>
       
      <div
        data-test="ec-metroline-item__header-cta"
      >
        <a
          href="#"
        >
          Edit
        </a>
      </div>
    </div>
     
    <div
      class="ec-metroline-item__main"
      data-test="ec-metroline-item__main"
    >
      <p>
         Lorem ipsum dolor sit amet, voluptates at cumque repudiandae atque quod voluptatum, aperiam dignissimos, vitae, neque mollitia repellat! 
      </p>
    </div>
     
    <div
      class="ec-metroline-item__footer"
      data-test="ec-metroline-item__footer"
    >
      <div
        class="ec-metroline-item__footer-cta"
        data-test="ec-metroline-item__footer-cta"
      >
        <a
          href="#"
        >
          Continue
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`EcMetrolineItem should render properly when is the last item and status is "active" 1`] = `
<div
  class="ec-metroline-item ec-metroline-item--is-active ec-metroline-item--is-last"
  data-test="ec-metroline-item"
>
  <div
    class="ec-metroline-item__status"
    data-test="ec-metroline-item__status"
  >
    <div
      class="ec-metroline-item__badge"
      data-test="ec-metroline-item__badge"
    >
      <span
        data-test="ec-metroline-item__index"
      >
        
        1
      
      </span>
    </div>
     
    <!---->
  </div>
   
  <div
    class="ec-metroline-item__content"
    data-test="ec-metroline-item__content"
  >
    <div
      class="ec-metroline-item__header"
      data-test="ec-metroline-item__header"
    >
      <div
        class="ec-metroline-item__header-heading"
        data-test="ec-metroline-item__header-heading"
      >
        <span>
          Payee information
        </span>
      </div>
       
      <!---->
       
      <!---->
    </div>
     
    <div
      class="ec-metroline-item__main"
      data-test="ec-metroline-item__main"
    >
      <p>
         Lorem ipsum dolor sit amet, voluptates at cumque repudiandae atque quod voluptatum, aperiam dignissimos, vitae, neque mollitia repellat! 
      </p>
    </div>
     
    <div
      class="ec-metroline-item__footer"
      data-test="ec-metroline-item__footer"
    >
      <div
        class="ec-metroline-item__footer-cta"
        data-test="ec-metroline-item__footer-cta"
      >
        <a
          href="#"
        >
          Continue
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`EcMetrolineItem should render properly when is the last item and status is "completed" 1`] = `
<div
  class="ec-metroline-item ec-metroline-item--is-active ec-metroline-item--is-last"
  data-test="ec-metroline-item"
>
  <div
    class="ec-metroline-item__status"
    data-test="ec-metroline-item__status"
  >
    <div
      class="ec-metroline-item__badge"
      data-test="ec-metroline-item__badge"
    >
      <svg
        class="ec-icon"
        data-test="ec-metroline-item__completed-icon"
        height="14"
        width="14"
      >
        <use
          xlink:href="#ec-simple-check"
        />
      </svg>
    </div>
     
    <!---->
  </div>
   
  <div
    class="ec-metroline-item__content"
    data-test="ec-metroline-item__content"
  >
    <div
      class="ec-metroline-item__header"
      data-test="ec-metroline-item__header"
    >
      <div
        class="ec-metroline-item__header-heading"
        data-test="ec-metroline-item__header-heading"
      >
        <span>
          Payee information
        </span>
      </div>
       
      <div
        class="ec-metroline-item__header-sub-heading"
        data-test="ec-metroline-item__header-sub-heading"
      >
        <span>
          Amount fully allocated
        </span>
      </div>
       
      <div
        data-test="ec-metroline-item__header-cta"
      >
        <a
          href="#"
        >
          Edit
        </a>
      </div>
    </div>
     
    <div
      class="ec-metroline-item__main"
      data-test="ec-metroline-item__main"
    >
      <p>
         Lorem ipsum dolor sit amet, voluptates at cumque repudiandae atque quod voluptatum, aperiam dignissimos, vitae, neque mollitia repellat! 
      </p>
    </div>
     
    <div
      class="ec-metroline-item__footer"
      data-test="ec-metroline-item__footer"
    >
      <div
        class="ec-metroline-item__footer-cta"
        data-test="ec-metroline-item__footer-cta"
      >
        <a
          href="#"
        >
          Continue
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`EcMetrolineItem should render properly when is the last item and status is "next" 1`] = `
<div
  class="ec-metroline-item ec-metroline-item--is-last"
  data-test="ec-metroline-item"
>
  <div
    class="ec-metroline-item__status"
    data-test="ec-metroline-item__status"
  >
    <div
      class="ec-metroline-item__badge"
      data-test="ec-metroline-item__badge"
    >
      <span
        data-test="ec-metroline-item__index"
      >
        
        1
      
      </span>
    </div>
     
    <!---->
  </div>
   
  <div
    class="ec-metroline-item__content"
    data-test="ec-metroline-item__content"
  >
    <div
      class="ec-metroline-item__header"
      data-test="ec-metroline-item__header"
    >
      <div
        class="ec-metroline-item__header-heading"
        data-test="ec-metroline-item__header-heading"
      >
        <span>
          Payee information
        </span>
      </div>
       
      <!---->
       
      <!---->
    </div>
     
    <!---->
     
    <!---->
  </div>
</div>
`;

exports[`EcMetrolineItem should render properly when status is "active" 1`] = `
<div
  class="ec-metroline-item ec-metroline-item--is-active"
  data-test="ec-metroline-item"
>
  <div
    class="ec-metroline-item__status"
    data-test="ec-metroline-item__status"
  >
    <div
      class="ec-metroline-item__badge"
      data-test="ec-metroline-item__badge"
    >
      <span
        data-test="ec-metroline-item__index"
      >
        
        1
      
      </span>
    </div>
     
    <div
      class="ec-metroline-item__status-bar"
      data-test="ec-metroline-item__status-bar"
    />
  </div>
   
  <div
    class="ec-metroline-item__content"
    data-test="ec-metroline-item__content"
  >
    <div
      class="ec-metroline-item__header"
      data-test="ec-metroline-item__header"
    >
      <div
        class="ec-metroline-item__header-heading"
        data-test="ec-metroline-item__header-heading"
      >
        <span>
          Payee information
        </span>
      </div>
       
      <!---->
       
      <!---->
    </div>
     
    <div
      class="ec-metroline-item__main"
      data-test="ec-metroline-item__main"
    >
      <p>
         Lorem ipsum dolor sit amet, voluptates at cumque repudiandae atque quod voluptatum, aperiam dignissimos, vitae, neque mollitia repellat! 
      </p>
    </div>
     
    <div
      class="ec-metroline-item__footer"
      data-test="ec-metroline-item__footer"
    >
      <div
        class="ec-metroline-item__footer-cta"
        data-test="ec-metroline-item__footer-cta"
      >
        <a
          href="#"
        >
          Continue
        </a>
      </div>
    </div>
  </div>
</div>
`;
